.github/scripts/workflow-monitor.py

```diff
@@ -32,44 +35,71 @@
+
+    try:
+        print("Sending startup message")
+        issue_post(ci_env['PERSONAL_ACCESS_TOKEN'], f"Workflow monitor started for CI run: {ci_env['GITHUB_RUN_ID']}")
+
+        print("Beginning polling loop")
+        while True:
+            time.sleep(POLLING_INTERVAL_SECONDS)
+
+            res = requests.get(gha_workflow_api_url, headers=get_header(ci_env['PERSONAL_ACCESS_TOKEN']))
+            if res.status_code == 200:
+                consecutive_failures = 0
+                res_dict = res.json()
+                state_status = res_dict["status"]
+                state_concl = res_dict["conclusion"]
+
+                print(f"Workflow {ci_env['GITHUB_RUN_ID']} status: {state_status} {state_concl}")
+
+                # check that select instances are terminated on time
+                platform_lib.check_and_terminate_select_instances(45, ci_env['GITHUB_RUN_ID'])
+
+                if state_status in ['completed']:
+                    if state_concl in TERMINATE_STATES:
+                        platform_lib.terminate_instances(ci_env['PERSONAL_ACCESS_TOKEN'], ci_env['GITHUB_RUN_ID'])
+                        return
+                    elif state_concl in STOP_STATES:
+                        platform_lib.stop_instances(ci_env['PERSONAL_ACCESS_TOKEN'], ci_env['GITHUB_RUN_ID'])
+                        return
+                    elif state_concl not in NOP_STATES:
+                        print(f"Unexpected Workflow State On Completed: {state_concl}")
+                        raise ValueError
+                elif state_status not in ['in_progress', 'queued', 'waiting', 'requested']:
+                    print(f"Unexpected Workflow State: {state_status}")
                     raise ValueError
-            elif state_status not in ['in_progress', 'queued', 'waiting', 'requested']:
-                print("Unexpected Workflow State: {}".format(state_status))
-                raise ValueError
 
-        else:
-            print("HTTP GET error: {}. Retrying.".format(res.json()))
-            consecutive_failures = consecutive_failures + 1
-            if consecutive_failures == QUERY_FAILURE_THRESHOLD:
-                platform_lib.stop_instances(gha_ci_personal_token, workflow_id)
-                exit(1)
+            else:
+                print(f"HTTP GET error: {res.json()}. Retrying.")
+                consecutive_failures = consecutive_failures + 1
+                if consecutive_failures == QUERY_FAILURE_THRESHOLD:
+                    print("Consecutive HTTP GET errors. Terminating and exiting.")
+                    platform_lib.terminate_instances(ci_env['PERSONAL_ACCESS_TOKEN'], ci_env['GITHUB_RUN_ID'])
+                    raise ValueError
+    except BaseException as e:
+        post_str  = f"Something went wrong in the workflow monitor for CI run {ci_env['GITHUB_RUN_ID']}. Verify CI instances are terminated properly. Must be checked before submitting the PR.\n\n"
```

Comment:
This is a tradeoff on what we think will fail. If the message posting fails, then this will skip the `terminate_instances` steps (since the post will throw an exception). However, if the initial post is moved after the `terminate_inst` step and the `terminate_inst` step fails then the post will never get sent. In either case, the "cull instances" cron job should catch both.